From ab7931b0efdd792f21551dea25a23e68c6f22b44 Mon Sep 17 00:00:00 2001
From: Pavan Naregundi <pnaregundi@marvell.com>
Date: Thu, 7 Mar 2024 06:03:42 +0000
Subject: sonic-installer fixes

Patch involves following fixes:
* Fixes config save and restore across sonic to sonic install.
* Set proper second image during.
* Fix reset of boot_once
* Fixes boot menu print.

Signed-off-by: Pavan Naregundi <pnaregundi@marvell.com>
---
 platform/marvell-arm64/platform.conf | 39 ++++++++++++++++++----------
 1 file changed, 25 insertions(+), 14 deletions(-)

diff --git a/platform/marvell-arm64/platform.conf b/platform/marvell-arm64/platform.conf
index 8a78164d9..c9f157ecf 100644
--- a/platform/marvell-arm64/platform.conf
+++ b/platform/marvell-arm64/platform.conf
@@ -181,12 +181,23 @@ prepare_boot_menu() {
         fit_name_old=""
         sonic_version_2="None"
     else
-        image_dir_old=$(fw_printenv -n image_dir || true)
-        image_name_old=$(fw_printenv -n image_name || true)
-        initrd_name_old=$(fw_printenv -n initrd_name || true)
-        fdt_name_old=$(fw_printenv -n fdt_name || true)
-        fit_name_old=$(fw_printenv -n fit_name || true)
-        sonic_version_2=$(fw_printenv -n sonic_version_1 || true)
+        CURR_SONIC_IMAGE="$(sonic-installer list | grep "Current: " | cut -f2 -d' ')"
+        FIRST_SONIC_IMAGE="$(fw_printenv sonic_version_1 | cut -f2 -d'=')"
+        if [ "$CURR_SONIC_IMAGE" = "$FIRST_SONIC_IMAGE" ]; then
+            image_dir_old=$(fw_printenv -n image_dir || true)
+            image_name_old=$(fw_printenv -n image_name || true)
+            initrd_name_old=$(fw_printenv -n initrd_name || true)
+            fdt_name_old=$(fw_printenv -n fdt_name || true)
+            fit_name_old=$(fw_printenv -n fit_name || true)
+            sonic_version_2=$(fw_printenv -n sonic_version_1 || true)
+        else
+            image_dir_old=$(fw_printenv -n image_dir_old || true)
+            image_name_old=$(fw_printenv -n image_name_old || true)
+            initrd_name_old=$(fw_printenv -n initrd_name_old || true)
+            fdt_name_old=$(fw_printenv -n fdt_name_old || true)
+            fit_name_old=$(fw_printenv -n fit_name_old || true)
+            sonic_version_2=$(fw_printenv -n sonic_version_2 || true)
+        fi
     fi
 
     # Set boot variables
@@ -206,12 +217,12 @@ prepare_boot_menu() {
     BOOT2='echo " > Boot2: $sonic_version_2 - run sonic_image_2";echo;'
     BOOT3='echo " > Boot3: ONIE - run onie_boot";echo;'
     BORDER='echo "---------------------------------------------------";echo;'
-    fw_setenv ${FW_ARG} print_menu $BORDER $BOOT1 $BOOT2 $BOOT3 $BORDER > /dev/null
+    fw_setenv ${FW_ARG} print_menu "$BORDER $BOOT1 $BOOT2 $BOOT3 $BORDER" > /dev/null
 
     fw_setenv ${FW_ARG} linuxargs "net.ifnames=0 loopfstype=squashfs loop=$image_dir/$FILESYSTEM_SQUASHFS systemd.unified_cgroup_hierarchy=0 varlog_size=$VAR_LOG ${ONIE_PLATFORM_EXTRA_CMDLINE_LINUX}" > /dev/null
     fw_setenv ${FW_ARG} linuxargs_old "net.ifnames=0 loopfstype=squashfs loop=$image_dir_old/$FILESYSTEM_SQUASHFS systemd.unified_cgroup_hierarchy=0 varlog_size=$VAR_LOG ${ONIE_PLATFORM_EXTRA_CMDLINE_LINUX}" > /dev/null
-    sonic_bootargs_old='setenv bootargs root='$demo_dev' rw rootwait rootfstype=ext4 panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs_old}'
-    fw_setenv ${FW_ARG} sonic_bootargs_old $sonic_bootargs_old > /dev/null || true
+    sonic_bootargs_old='setenv bootargs root='$demo_dev' rw rootwait panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs_old}'
+    fw_setenv ${FW_ARG} sonic_bootargs_old "$sonic_bootargs_old" > /dev/null || true
     sonic_boot_load_old=$(fw_printenv -n sonic_boot_load || true)
     old_str="_old"
     fw_setenv ${FW_ARG} sonic_boot_load_old "$sonic_boot_load_old$old_str" > /dev/null || true
@@ -225,14 +236,14 @@ prepare_boot_menu() {
     fw_setenv ${FW_ARG} sonic_boot_load "$MMC_LOAD" > /dev/null
     SONIC_BOOT_CMD='run sonic_bootargs; run sonic_boot_load; bootm $fit_addr${fit_conf_name}'
     SONIC_BOOT_CMD_OLD='run sonic_bootargs_old; run sonic_boot_load_old; bootm $fit_addr${fit_conf_name}'
-    BOOTARGS='setenv bootargs root='$demo_dev' rw rootwait rootfstype=ext4 panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs}'
-    fw_setenv ${FW_ARG} sonic_bootargs $BOOTARGS > /dev/null
-    fw_setenv ${FW_ARG} sonic_bootcmd $SONIC_BOOT_CMD > /dev/null
-    fw_setenv ${FW_ARG} sonic_image_2 $SONIC_BOOT_CMD_OLD > /dev/null
+    BOOTARGS='setenv bootargs root='$demo_dev' rw rootwait panic=1 console=ttyS0,${baudrate} ${othbootargs} ${mtdparts} ${linuxargs}'
+    fw_setenv ${FW_ARG} sonic_bootargs "$BOOTARGS" > /dev/null
+    fw_setenv ${FW_ARG} sonic_image_2 "$SONIC_BOOT_CMD_OLD" > /dev/null
     fw_setenv ${FW_ARG} sonic_image_1 "$SONIC_BOOT_CMD" > /dev/null
     fw_setenv ${FW_ARG} boot_next  'run sonic_image_1'> /dev/null
-    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && run boot_once; run boot_next' > /dev/null
+    fw_setenv ${FW_ARG} bootcmd 'run print_menu; usb start; test -n "$boot_once" && setenv boot_once "" && saveenv && run boot_once; run boot_next' > /dev/null
 
+    echo "Installed SONiC base image SONiC-OS successfully"
 }
 
 bootloader_menu_config() {
-- 
2.25.1

